cpp/src/sampling/random_walks_impl.cuh

```diff
@@ -70,21 +68,132 @@ struct sample_edges_op_t {
   }
 };
 
+template <typename vertex_t, typename bias_t>
+struct biased_random_walk_e_bias_op_t {
+
+  __device__ bias_t
+  operator()(vertex_t, vertex_t, bias_t src_out_weight_sum, thrust::nullopt_t, bias_t weight) const
+  {
+    return weight / src_out_weight_sum;
+  }
+};
+
+template <typename vertex_t, typename weight_t>
+struct biased_sample_edges_op_t {
+  __device__ thrust::tuple<vertex_t, weight_t>
+  operator()(vertex_t, vertex_t dst, weight_t, thrust::nullopt_t, weight_t weight) const
+  {
+    return thrust::make_tuple(dst, weight);
+  }
+};
+
+template <typename vertex_t, typename bias_t, typename weight_t>
+struct node2vec_random_walk_e_bias_op_t {
+  bias_t p_{};
+  bias_t q_{};
+  raft::device_span<size_t const> intersection_offsets_{};
+  raft::device_span<vertex_t const> intersection_indices_{};
+  raft::device_span<vertex_t const> current_vertices_{};
+  raft::device_span<vertex_t const> prev_vertices_{};
+
+  // Unweighted Bias Operator
+  template <typename W = weight_t>
+  __device__ std::enable_if_t<std::is_same_v<W, void>, bias_t> operator()(
+        thrust::tuple<vertex_t, vertex_t> tagged_src,
+        vertex_t dst,
+        thrust::nullopt_t,
+        thrust::nullopt_t,
+        thrust::nullopt_t) const
+  {
+    //  Check tag (prev vert) for destination
+    if(dst == thrust::get<1>(tagged_src)){
+      return 1.0 / p_;
+    }
+    //  Search zipped vertices for tagged src
+    auto lower_itr = thrust::lower_bound(
+            thrust::seq, 
+            thrust::make_zip_iterator(current_vertices_.begin(), prev_vertices_.begin()),
+            thrust::make_zip_iterator(current_vertices_.end(), prev_vertices_.end()),
+            tagged_src);
+    auto low_idx = thrust::distance(thrust::make_zip_iterator(current_vertices_.begin(), 
+                                                              prev_vertices_.begin()), 
+                                    lower_itr);
+    auto intersection_index_first = intersection_indices_.begin() + intersection_offsets_[low_idx];
+    auto intersection_index_last = intersection_indices_.begin() + intersection_offsets_[low_idx + 1];
+    auto itr = thrust::lower_bound(thrust::seq, intersection_index_first, intersection_index_last, dst);
+    return (itr != intersection_index_last && *itr == dst) ? 1.0 : 1.0 / q_;
+  }
+
+  //  Weighted Bias Operator
+  template <typename W = weight_t>
+  __device__ std::enable_if_t<!std::is_same_v<W, void>, bias_t> operator()(
+        thrust::tuple<vertex_t, vertex_t> tagged_src,
+        vertex_t dst,
+        thrust::nullopt_t,
+        thrust::nullopt_t,
+        W w) const
```

Comment:
Where are we using `w`? If this is not used, `W w`=> `W` will be sufficient (and also make it clear that `w` is not used).